src/metric/auc.cu

```diff
@@ -0,0 +1,547 @@
+/*!
+ * Copyright 2021 by XGBoost Contributors
+ */
+#include <thrust/scan.h>
+#include <cub/cub.cuh>
+#include <cassert>
+#include <limits>
+#include <memory>
+#include <utility>
+#include <tuple>
+
+#include "rabit/rabit.h"
+#include "xgboost/span.h"
+#include "xgboost/data.h"
+#include "auc.h"
+#include "../common/device_helpers.cuh"
+#include "../common/ranking_utils.cuh"
+
+namespace xgboost {
+namespace metric {
+namespace {
+template <typename T>
+class Discard : public thrust::discard_iterator<T>  {
+ public:
+  using value_type = T;  // NOLINT
+};
+
+struct GetWeightOp {
+  common::Span<float const> weights;
+  common::Span<size_t const> sorted_idx;
+
+  __device__ float operator()(size_t i) const {
+    return weights.empty() ? 1.0f : weights[sorted_idx[i]];
+  }
+};
+}  // namespace
+
+/**
+ * A cache to GPU data to avoid reallocating memory.
+ */
+struct DeviceAUCCache {
```

Comment:
How necessary are these cached items? Can you use dh::TemporaryArray to avoid caching? If I remember correctly caching_device_vector currently has synchronisation overhead due to upstream changes in thrust.

Reply:
> has synchronisation overhead

Yes, happens in `thrust::fill_n`.  I will try to use temp array instead.

> How necessary are these cached items?

Just trying to avoid `cudaMalloc` for every iteration.

Reply:
Swapped some caching device vectors with temporary arrays.